### PR TITLE
Adding a Cassandra upgrade test

### PR DESCRIPTION
**Pull request by maisem**
Tests for #38336

### CODE REVIEW

**Comment by kargakis** [comment]
@maisem is this testing something more not covered from the other statefulset upgrade test you are adding?

@kubernetes/sig-apps-pr-reviews

**Comment by kargakis** [comment]
@kubernetes/sig-testing-misc @kubernetes/sig-api-machinery-misc got any opinions on pulling in dependencies in the main repo just for testing?

**Comment by spiffxp** [comment]
in the abstract, not a fan, and think this should be something in the "examples" repo instead, but I'm not 100% certain

**Comment by timothysc** [comment]
remove, this does not belong here.  --> Helm or examples.

**Comment by maisem** [comment]
@kargakis @timothysc @spiffxp the idea behind these tests to verify that really applications behave correctly during upgrades. e.g. this test checks how cassandra behaves.

There are two aspects to this test (and we can have the mysql test do something similar as well)
1. We have a test image which handles communication with cassandra using gocql.
2. The actual test which communicates with the service

If the primary concern is over pulling dependencies, we can remove the image building aspect from this repo and move it somewhere else (e.g. https://github.com/kubernetes/examples). That would split the build stage but would still make it possible to reuse the upgrade framework.

WDYT?

**Comment by kargakis** [comment]
If all we want to test is application availablility during node upgrades can't we just use an existing test image like nginx and call it a day? Why do we have to test specifically Cassandra and MySQL? In all cases we should be able to verify that the container is up and running so the extra dependencies seem redundant to me.

**Comment by maisem** [comment]
Using nginx as a StatefulSet doesn't help because nginx isn't stateful as it doesn't rely on the guarantees that are provided by the controller.
Testing the StatefulSet capability e.g. stable pod identities, feels more like a unit test.

Cassandra provides us a test where we require quorum to read/write and uses PVCs.
MySQL provides us a test where we have a single writer with multiple readers and also uses PVCs.

Writing a custom application that actually relies on these capabilities is probably going to be highly buggy. Do you have an alternative suggestion? The goal here is to make sure that we are actually getting the correct signals.

**Comment by kargakis** [comment]
I am not sure that nginx being stateless really matters in this case. If we specifically want to test quorum reads, I am fine with it but from the k8s side doesn't this always translate into having a minimum number of pods running at all times?

**Comment by mml** [comment]
We're really testing that the reccomendations we give for running stateful applications, combined with a real stateful application and the real running k8s code, work together to provide the availability we think they should.  There are a lot of things to get wrong (I'm working on a doc that explains this): the logic in the node upgrade code, the PDB specs, the PDB controller, the evict handler, pod lifecycle stuff like signal handling and grace period, readiness and liveness probes correctly configured.

You are right that we can test most of these things separately.  We do.  But we want an end to end test that tells us in no uncertain terms when a typical application wouldn't get the results we expect just by following our recommendations.

That is what this test does.

**Comment by timothysc** [comment]
> the idea behind these tests to verify that real applications behave correctly during upgrades. e.g. this test checks how cassandra behaves.

I like the idea, but this is the wrong repository to be doing this in.  We've had numerous discussions that application level evaluation should occur our of core kubernetes repository.

**Comment by mml** [comment]
> I like the idea, but this is the wrong repository to be doing this in.

The more we talk about this, the more I agree.  These application level tests belong somewhere else.  Can we agree to do that move post-1.7?  For now, this looks like the only place where we have the infrastructure to easily check these in and get them continuously executed.

If it helps, I'm pretty sure @maisem would want to move the Cassandra image code someplace before merging this.  Probably to examples.

**Comment by maisem** [comment]
@kargakis @timothysc WDYT?

**Comment by kargakis** [comment]
> Can we agree to do that move post-1.7? For now, this looks like the only place where we have the infrastructure to easily check these in and get them continuously executed.

Is this a 1.7 blocker? Can we make the effort to enable testing in the examples repo? I would expect it to merely be a new stanza in the prow config, but @kubernetes/sig-testing-misc can confirm.

**Comment by mml** [comment]
Yes, this is for #38336, which is for 1.7.

**Comment by kargakis** [comment]
I am ok with merging but this needs a follow-up issue to investigate what needs tbd for moving these tests in the examples repo.

File: test/e2e/upgrades/cassandra.go
Comment (mml) [inline test/e2e/upgrades/cassandra.go]: It would be good to describe the overall test strategy in a comment block above the method.  That we are writing users in one thread and polling how many there are in another thread, and we're shooting for never seeing missing users and for a write success rate above X%.

For a naive reader, it is some reading between the lines to figure out that this is the goal.

**Review by mml — CHANGES_REQUESTED** [review]
Just need a comment.  LG otherwise.  Thanks @maisem.

**Comment by mml** [comment]
> I am ok with merging but this needs a follow-up issue to investigate what needs tbd for moving these tests in the examples repo.

#47930

**Comment by mml** [comment]
As a followup, kubernetes/examples doesn't currently have support for including vendored dependencies for the image build.  So that is going to be another hurdle.

I guess in a nutshell, the destination may not be examples, and the work involved might be more than we anticipated, but we will follow up in #47930 to find a suitable destination for both the images and the tests.

**Comment by timothysc** [comment]
TBH I think folks should talk with SIG-apps and perhaps look at their helm charts and consider tests there.

**Comment by mml** [comment]
@kargakis Also need approval.

**Comment by lavalamp** [comment]
I am really not excited about adding these dependencies for a test. Can the test image build be moved elsewhere? The test image really shouldn't ever need to change, right?

**Comment by lavalamp** [comment]
To unblock yourself, you could make a personal repo to hold the test image code, and figure out a better home later.

We can't keep adding tech debt to the main repo--I would consider these deps tech debt if we all think they need to go somewhere else later. Esp. considering that 4 or 5 people in this thread all had the same reaction.

It's 1000x easier to keep tech debt out in the first place than to move it out at a later point in time.

**Comment by kargakis** [comment]
Agreed with @lavalamp

**Comment by foxish** [comment]
I recently setup https://k8s-testgrid.appspot.com/kubernetes-apps which tests certain important helm charts against head k8s. I can imagine a similar testsuite for these application upgrade tests as well which exists outside the repo.

**Comment by maisem** [comment]
Moved the images to kubernetes/examples#19

**Comment by mml** [comment]
@lavalamp PTAL

**Comment by mml** [comment]
@maisem I just realized this never got merged (!).  Can you get these tests passing so we can get this in?

**Pull request merged**
